server/chat-plugins/abuse-monitor.ts

```diff
@@ -666,6 +720,101 @@ export const commands: Chat.ChatCommands = {
 		},
+		ap: 'addpunishment',
+		addpunishment(target, room, user) {
+			checkAccess(this);
+			if (!toID(target)) return this.parse(`/help am`);
+			const targets = target.split(',').map(f => f.trim());
+			const punishment: Partial<PunishmentSettings> = {};
+			for (const cur of targets) {
+				let [key, value] = Utils.splitFirst(cur, '=').map(f => f.trim());
+				key = toID(key);
+				switch (key) {
+				case 'punishment': case 'p':
+					if (punishment.punishment) {
+						return this.errorReply(`Duplicate punishment values.`);
+					}
+					value = toID(value).toUpperCase();
+					if (!PUNISHMENTS.includes(value)) {
+						return this.errorReply(`Invalid punishment: ${value}. Valid punishments: ${PUNISHMENTS.join(', ')}.`);
+					}
+					punishment.punishment = value;
+					break;
+				case 'count': case 'num': case 'c':
+					if (punishment.count) {
+						return this.errorReply(`Duplicate count values.`);
+					}
+					const num = parseInt(value);
+					if (isNaN(num)) {
+						return this.errorReply(`Invalid count '${value}'. Must be a number.`);
+					}
+					punishment.count = num;
+					break;
+				case 'type': case 't':
+					if (punishment.type) {
+						return this.errorReply(`Duplicate type values.`);
+					}
+					value = value.replace(/\s/g, '_').toUpperCase();
+					if (!ATTRIBUTES[value as keyof typeof ATTRIBUTES]) {
+						return this.errorReply(
+							`Invalid attribute: ${value}. ` +
+							`Valid attributes: ${Object.keys(ATTRIBUTES).join(', ')}.`
+						);
+					}
+					punishment.type = value;
+					break;
+				case 'certainty': case 'ct':
+					if (punishment.certainty) {
+						return this.errorReply(`Duplicate certainty values.`);
+					}
+					const certainty = parseFloat(value);
+					if (isNaN(certainty) || certainty > 1 || certainty < 0) {
+						return this.errorReply(`Invalid certainty '${value}'. Must be a number above 0 and below 1.`);
+					}
+					punishment.certainty = certainty;
+					break;
+				default:
+					this.errorReply(`Invalid key:  ${key}`);
+					return this.parse(`/help am`);
+				}
+			}
+			if (!punishment.punishment) {
+				return this.errorReply(`A punishment type must be specified.`);
+			}
+			for (const [i, p] of settings.punishments.entries()) {
+				let matches = 0;
+				for (const k in p) {
+					if (p[k as keyof PunishmentSettings] === punishment[k as keyof PunishmentSettings]) matches++;
+				}
+				if (matches === Object.keys(p).length) {
+					return this.errorReply(`This punishment is already stored at ${i + 1}.`);
+				}
+			}
+			settings.punishments.push(punishment as PunishmentSettings);
+			saveSettings();
+			this.privateGlobalModAction(`${user.name} added a ${punishment.punishment} abuse-monitor punishment.`);
+			const str = Object.keys(punishment).map(f => `${f}: ${punishment[f as keyof PunishmentSettings]}`).join(', ');
+			this.stafflog(`Info: ${str}`);
```

Comment:
What's this for?

Reply:
It's to log what it looks like after the change (for when one is reading chatlogs). I just didn't want to figure out how to prettify it into a mod action message, so I logged the data instead.

Reply:
Just posting in staff room with `Info:` as the only way to tell what the information isn't very helpful - don't you already have an entire room dedicated to the abuse monitor?

Reply:
Yes, but this is logged directly after the action stating it happened - so one can see that this is the info on the added punishment.